### ISSUE TITLE
Piecewise and evaluate(False). Version of sympy 1.7.1

### ISSUE THREAD

**Issue by satels** [sympy/sympy#20715]
```
>>> from sympy import *
>>> from sympy import __version__
>>> __version__
'1.7.1'
>>> from sympy.abc import x
>>> Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))
Piecewise((-1, x < 2), (2*x - 1, x <= 3), (1, True))
>>> with evaluate(False): Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))
... 
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'positive'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'finite'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'finite'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'finite'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'infinite'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'imaginary'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'imaginary'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'imaginary'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sympy/core/assumptions.py", line 454, in getit
    return self._assumptions[fact]
KeyError: 'imaginary'

During handling of the above exception, another exception occurred:
```

Also in 1.6.2 version:
```
>>> from sympy import *
>>> from sympy import __version__
>>> __version__
'1.6.2'
>>> from sympy.abc import x
>>> Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))
Piecewise((-1, x < 2), (2*x - 1, x <= 3), (1, True))
>>> with evaluate(False): Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/sympy/functions/elementary/piecewise.py", line 136, in __new__
    r = cls.eval(*newargs)
  File "/usr/local/lib/python3.9/site-packages/sympy/functions/elementary/piecewise.py", line 187, in eval
    c = c.as_set().as_relational(x)
  File "/usr/local/lib/python3.9/site-packages/sympy/logic/boolalg.py", line 148, in as_set
    if periodicity(r, x) not in (0, None):
  File "/usr/local/lib/python3.9/site-packages/sympy/calculus/util.py", line 431, in periodicity
    f = simplify(f)
  File "/usr/local/lib/python3.9/site-packages/sympy/simplify/simplify.py", line 601, in simplify
    expr = Mul(*powsimp(expr).as_content_primitive())
  File "/usr/local/lib/python3.9/site-packages/sympy/core/power.py", line 1722, in as_content_primitive
    b = _keep_coeff(*b.as_content_primitive(radical=radical, clear=clear))
  File "/usr/local/lib/python3.9/site-packages/sympy/core/mul.py", line 1798, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/usr/local/lib/python3.9/site-packages/sympy/core/mul.py", line 1798, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/usr/local/lib/python3.9/site-packages/sympy/core/power.py", line 1747, in as_content_primitive
    h, t = b.as_content_primitive(radical=radical, clear=clear)  # h is positive
  File "/usr/local/lib/python3.9/site-packages/sympy/core/mul.py", line 1798, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/usr/local/lib/python3.9/site-packages/sympy/core/power.py", line 1723, in as_content_primitive
    ce, pe = e.as_content_primitive(radical=radical, clear=clear)
  File "/usr/local/lib/python3.9/site-packages/sympy/core/mul.py", line 1798, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/usr/local/lib/python3.9/site-packages/sympy/core/power.py", line 1741, in as_content_primitive
    iceh, r = divmod(ceh.p, ceh.q)
AttributeError: 'Mul' object has no attribute 'p'

```

**Comment by oscarbenjamin** [comment]
Part of this looks related to #20597 

I get different results each time I run this (on master):
```python
from sympy import *

x = Symbol('x')

with evaluate(False):
    Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))
```
Sometimes I get this which looks like #20597:
```console
$ python qwe.py 
Traceback (most recent call last):
  File "/Users/enojb/current/sympy/sympy/sympy/core/compatibility.py", line 301, in as_int
    return operator.index(n)
TypeError: 'Abs' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "qwe.py", line 6, in <module>
    Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))
  File "/Users/enojb/current/sympy/sympy/sympy/functions/elementary/piecewise.py", line 136, in __new__
    r = cls.eval(*newargs)
  File "/Users/enojb/current/sympy/sympy/sympy/functions/elementary/piecewise.py", line 187, in eval
    c = c.as_set().as_relational(x)
  File "/Users/enojb/current/sympy/sympy/sympy/logic/boolalg.py", line 160, in as_set
    if periodicity(r, x) not in (0, None):
  File "/Users/enojb/current/sympy/sympy/sympy/calculus/util.py", line 432, in periodicity
    f = simplify(f)
  File "/Users/enojb/current/sympy/sympy/sympy/simplify/simplify.py", line 621, in simplify
    _e = cancel(expr)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polytools.py", line 6668, in cancel
    f = factor_terms(f, radical=True)
  File "/Users/enojb/current/sympy/sympy/sympy/core/exprtools.py", line 1274, in factor_terms
    return do(expr)
  File "/Users/enojb/current/sympy/sympy/sympy/core/exprtools.py", line 1264, in do
    p = gcd_terms(p,
  File "/Users/enojb/current/sympy/sympy/sympy/core/exprtools.py", line 1077, in gcd_terms
    cont, numer, denom = _gcd_terms(terms, isprimitive, fraction)
  File "/Users/enojb/current/sympy/sympy/sympy/core/exprtools.py", line 959, in _gcd_terms
    cont = cont.gcd(term)
  File "/Users/enojb/current/sympy/sympy/sympy/core/exprtools.py", line 888, in gcd
    return Term(self.coeff.gcd(other.coeff),
  File "/Users/enojb/current/sympy/sympy/sympy/core/decorators.py", line 95, in __sympifyit_wrapper
    return func(a, b)
  File "/Users/enojb/current/sympy/sympy/sympy/core/numbers.py", line 1990, in gcd
    return Rational(
  File "/Users/enojb/current/sympy/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/sympy/sympy/core/numbers.py", line 1653, in __new__
    gcd = igcd(abs(p), q)
  File "/Users/enojb/current/sympy/sympy/sympy/core/numbers.py", line 254, in igcd
    args_temp = [abs(as_int(i)) for i in args]
  File "/Users/enojb/current/sympy/sympy/sympy/core/numbers.py", line 254, in <listcomp>
    args_temp = [abs(as_int(i)) for i in args]
  File "/Users/enojb/current/sympy/sympy/sympy/core/compatibility.py", line 303, in as_int
    raise ValueError('%s is not an integer' % (n,))
ValueError: Abs(1*1) is not an integer
```

Other times I get a recursion error. It seems that the recursion goes wrong with
```
    _ask(pk, obj)
  File "/Users/enojb/current/sympy/sympy/sympy/core/assumptions.py", line 513, in _ask
    _ask(pk, obj)
  File "/Users/enojb/current/sympy/sympy/sympy/core/assumptions.py", line 513, in _ask
    _ask(pk, obj)
  [Previous line repeated 5 more times]
  File "/Users/enojb/current/sympy/sympy/sympy/core/assumptions.py", line 501, in _ask
    a = evaluate(obj)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 853, in _eval_is_negative
    extended_negative = self.is_extended_negative
  File "/Users/enojb/current/sympy/sympy/sympy/core/assumptions.py", line 458, in getit
    return _ask(fact, self)
  File "/Users/enojb/current/sympy/sympy/sympy/core/assumptions.py", line 501, in _ask
File "/Users/enojb/current/sympy/sympy/sympy/core/assumptions.py", line 513, in _ask
    _ask(pk, obj)
  [Previous line repeated 5 more times]
  File "/Users/enojb/current/sympy/sympy/sympy/core/assumptions.py", line 501, in _ask
    a = evaluate(obj)
  File "/Users/enojb/current/sympy/sympy/sympy/core/add.py", line 712, in _eval_is_extended_nonnegative
    if not self.is_number:
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 449, in is_number
    return all(obj.is_number for obj in self.args)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 449, in <genexpr>
    return all(obj.is_number for obj in self.args)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 449, in is_number
    return all(obj.is_number for obj in self.args)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 449, in <genexpr>
    return all(obj.is_number for obj in self.args)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 449, in is_number
    return all(obj.is_number for obj in self.args)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 449, in <genexpr>
    return all(obj.is_number for obj in self.args)
```
In that case looking through the debugger self is something like:
```
(1/(1/(1/(1/(1/(1/(1/(1/(1/(1/(1/(1/_y))))))))))), -1)
```
(It get's longer as you go up the call stack).

I think that the cuplrit is here:
https://github.com/sympy/sympy/blob/17cf0f5ad8c2535e32009869220654e7fd25865f/sympy/core/power.py#L711-L712
Under evaluate=False that leads to an infinite recursion.

**Comment by oscarbenjamin** [comment]
The fix might be as simple as:
```diff
diff --git a/sympy/core/power.py b/sympy/core/power.py
index 85d1d2e892..8d5dddf8b4 100644
--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -708,9 +708,6 @@ def _eval_is_imaginary(self):
             if isodd is not None:
                 return isodd
 
-        if self.exp.is_negative:
-            return (1/self).is_imaginary
-
     def _eval_is_odd(self):
         if self.exp.is_integer:
             if self.exp.is_positive:
```
That passes all the core tests.

With that if I rerun the script I get the same error from #20597 but also sometimes the other one you showed:
```
$ python qwe.py 
Traceback (most recent call last):
  File "qwe.py", line 6, in <module>
    Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))
  File "/Users/enojb/current/sympy/sympy/sympy/functions/elementary/piecewise.py", line 136, in __new__
    r = cls.eval(*newargs)
  File "/Users/enojb/current/sympy/sympy/sympy/functions/elementary/piecewise.py", line 187, in eval
    c = c.as_set().as_relational(x)
  File "/Users/enojb/current/sympy/sympy/sympy/logic/boolalg.py", line 160, in as_set
    if periodicity(r, x) not in (0, None):
  File "/Users/enojb/current/sympy/sympy/sympy/calculus/util.py", line 432, in periodicity
    f = simplify(f)
  File "/Users/enojb/current/sympy/sympy/sympy/simplify/simplify.py", line 620, in simplify
    expr = Mul(*powsimp(expr).as_content_primitive())
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1672, in as_content_primitive
    b = _keep_coeff(*b.as_content_primitive(radical=radical, clear=clear))
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1672, in as_content_primitive
    b = _keep_coeff(*b.as_content_primitive(radical=radical, clear=clear))
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1672, in as_content_primitive
    b = _keep_coeff(*b.as_content_primitive(radical=radical, clear=clear))
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1672, in as_content_primitive
    b = _keep_coeff(*b.as_content_primitive(radical=radical, clear=clear))
  File "/Users/enojb/current/sympy/sympy/sympy/core/mul.py", line 1898, in as_content_primitive
    c, p = a.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1673, in as_content_primitive
    ce, pe = e.as_content_primitive(radical=radical, clear=clear)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1691, in as_content_primitive
    iceh, r = divmod(ceh.p, ceh.q)
AttributeError: 'Mul' object has no attribute 'p'
```
I think that has the same underlying cause which is that expand has not resolved the unexpanded mul.

I think that if the issue with `Pow._eval_is_imaginary` was fixed then this might be a duplicate of #20597.

I've just discovered another more fundamental problem though:
```
$ cat rty.py 
from sympy import *

x = S(-1)

with evaluate(False):
    print(x.is_negative)
$ python rty.py 
False
```
Somehow `evaluate(False)` has messed with the assumptions system quite badly here...

**Comment by oscarbenjamin** [comment]
Another problem is this:
```
$ cat rty.py 
from sympy import *

x = S(-1)

with evaluate(False):
    print(x.evalf(2).as_real_imag())

with evaluate(True):
    print(x.evalf(2).as_real_imag())
$ python rty.py 
(re(-1.0), im(-1.0))
(-1.0, 0)
```
That causes a problem here:
https://github.com/sympy/sympy/blob/17cf0f5ad8c2535e32009869220654e7fd25865f/sympy/core/expr.py#L881-L883

**Comment by oscarbenjamin** [comment]
I think that the problems in the assumptions system can be fixed with something like this:
```diff
diff --git a/sympy/core/expr.py b/sympy/core/expr.py
index f985bba2db..a93f203ac4 100644
--- a/sympy/core/expr.py
+++ b/sympy/core/expr.py
@@ -857,6 +857,7 @@ def _eval_is_negative(self):
             return False
 
     def _eval_is_extended_positive_negative(self, positive):
+        from sympy.core.numbers import pure_complex
         from sympy.polys.numberfields import minimal_polynomial
         from sympy.polys.polyerrors import NotAlgebraic
         if self.is_number:
@@ -878,9 +879,15 @@ def _eval_is_extended_positive_negative(self, positive):
             if n2 is S.NaN:
                 return None
 
-            r, i = self.evalf(2).as_real_imag()
-            if not i.is_Number or not r.is_Number:
-                return False
+            f = self.evalf(2)
+            if f.is_Float:
+                match = f, S.Zero
+            else:
+                match = pure_complex(f)
+            if match is not None:
+                r, i = match
+                if not (i.is_Number and r.is_Number):
+                    return False
             if r._prec != 1 and i._prec != 1:
                 return bool(not i and ((r > 0) if positive else (r < 0)))
             elif r._prec == 1 and (not i or i._prec == 1) and \
diff --git a/sympy/core/power.py b/sympy/core/power.py
index 85d1d2e892..8d5dddf8b4 100644
--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -708,9 +708,6 @@ def _eval_is_imaginary(self):
             if isodd is not None:
                 return isodd
 
-        if self.exp.is_negative:
-            return (1/self).is_imaginary
-
     def _eval_is_odd(self):
         if self.exp.is_integer:
             if self.exp.is_positive:
```
With that I consistently get the
```
AttributeError: 'Mul' object has no attribute 'p'
```
error creating the Piecewise which I think is the same issue as #20597

**Comment by oscarbenjamin** [comment]
Okay, this diff passes all tests in core:
```diff
diff --git a/sympy/core/expr.py b/sympy/core/expr.py
index f985bba2db..1517356402 100644
--- a/sympy/core/expr.py
+++ b/sympy/core/expr.py
@@ -857,6 +857,7 @@ def _eval_is_negative(self):
             return False
 
     def _eval_is_extended_positive_negative(self, positive):
+        from sympy.core.numbers import pure_complex
         from sympy.polys.numberfields import minimal_polynomial
         from sympy.polys.polyerrors import NotAlgebraic
         if self.is_number:
@@ -878,8 +879,15 @@ def _eval_is_extended_positive_negative(self, positive):
             if n2 is S.NaN:
                 return None
 
-            r, i = self.evalf(2).as_real_imag()
-            if not i.is_Number or not r.is_Number:
+            f = self.evalf(2)
+            if f.is_Float:
+                match = f, S.Zero
+            else:
+                match = pure_complex(f)
+            if match is None:
+                return False
+            r, i = match
+            if not (i.is_Number and r.is_Number):
                 return False
             if r._prec != 1 and i._prec != 1:
                 return bool(not i and ((r > 0) if positive else (r < 0)))
diff --git a/sympy/core/power.py b/sympy/core/power.py
index 85d1d2e892..8d5dddf8b4 100644
--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -708,9 +708,6 @@ def _eval_is_imaginary(self):
             if isodd is not None:
                 return isodd
 
-        if self.exp.is_negative:
-            return (1/self).is_imaginary
-
     def _eval_is_odd(self):
         if self.exp.is_integer:
             if self.exp.is_positive:
```
That should be applied with tests like:
```
x = S(-1)

with evaluate(False):
    assert x.is_negative is True

f = Pow(x, -1)

with evaluate(False):
    assert f.is_imaginary is None
```
Those tests probably belong somewhere in `test_arit.py`. There should be appropriate test functions where they can be added along with a comment linking to this issue:
```
# https://github.com/sympy/sympy/issues/20715
```

**Comment by Mayank-gaur** [comment]
I am a beginner in the open source world and I want to work on this issue. Can someone please guide me how can i get started on this? (I already have sympy environment installed locally)

**Comment by hyadav2k** [comment]
@oscarbenjamin @satels, Should the test cases be added in `test_pow_im()` in `test_arit.py`?

**Comment by 0sidharth** [comment]
> Okay, this diff passes all tests in core:

I think this only fixes the `RecursionError`, I am still getting `AttributeError: 'Mul' object has no attribute 'p'` with the diff you have mentioned.

**Comment by 0sidharth** [comment]
> error creating the Piecewise which I think is the same issue as #20597

It is the same error message, but the PR that fixes that does not fix this issue. A different fix is still needed.

**Comment by oscarbenjamin** [comment]
After #20721 this is partially fixed.

Now it either gives one of these two errors:
```
AttributeError: 'Mul' object has no attribute 'p'
```
Or otherwise
```
ValueError: Abs(1*1) is not an integer
```
I think these are both the same as #20597

**Comment by AaronStiff** [comment]
As regards the ValueError:

At one point in the traceback, it tries to create a `Rational` where `p = 1*1` and `q = 1`. However, in doing this, it calls `igcd`, which seems like it's only supposed to accept integers. Is `Rational` supposed to be able to work with non-integer arguments? Because if so, it shouldn't rely on `igcd`.

**Comment by oscarbenjamin** [comment]
The problem is wherever it is that the unevaluated `1*1` comes from. Basically we shouldn't assume in core functions that we can multiply by `1` if we want them to work under `evaluate(False)`. I've made a little headway with:
```diff
diff --git a/sympy/simplify/powsimp.py b/sympy/simplify/powsimp.py
index b682413b60..975f325de5 100644
--- a/sympy/simplify/powsimp.py
+++ b/sympy/simplify/powsimp.py
@@ -131,7 +131,10 @@ def recurse(arg, **kwargs):
         coeff = S.One
         for term in expr.args:
             if term.is_Rational:
-                coeff *= term
+                if coeff is S.One:
+                    coeff = term
+                else:
+                    coeff *= term
                 continue
             if term.is_Pow:
                 term = _denest_pow(term)
diff --git a/sympy/simplify/simplify.py b/sympy/simplify/simplify.py
index 197c800ffa..033df0c6d7 100644
--- a/sympy/simplify/simplify.py
+++ b/sympy/simplify/simplify.py
@@ -616,8 +616,14 @@ def done(e):
         floats = True
         expr = nsimplify(expr, rational=True)
 
+    def _factor_content(expr):
+        content, primitive = expr.as_content_primitive()
+        if S.One not in (content, primitive):
+            expr = Mul(content, primitive)
+        return expr
+
     expr = bottom_up(expr, lambda w: getattr(w, 'normal', lambda: w)())
-    expr = Mul(*powsimp(expr).as_content_primitive())
+    expr = _factor_content(powsimp(expr))
     _e = cancel(expr)
     expr1 = shorter(_e, _mexpand(_e).cancel())  # issue 6829
     expr2 = shorter(together(expr, deep=True), together(expr1, deep=True))
```

Another problem is the fact that creating a `Piecewise` calls `simplify` at all (especially under `evaluate=False`). That should be fixed by checking the global evaluate option in `Piecewise.__new__`.

Even when `evaluate=True` we shouldn't use `as_set` here:
https://github.com/sympy/sympy/blob/c872229fdc9f481e456d234f812169f010a23cbe/sympy/functions/elementary/piecewise.py#L186-L189

That should be fixed but also there are lots of little bugs around I think to do with multiplying by `1` are creating new instances in the assumptions system that are contributing to this.

**Comment by oscarbenjamin** [comment]
> At one point in the traceback, it tries to create a `Rational` where `p = 1*1` and `q = 1`. However, in doing this, it calls `igcd`, which seems like it's only supposed to accept integers.
> Is `Rational` supposed to be able to work with non-integer arguments? Because if so, it shouldn't rely on `igcd`.

Rational should not accept arbitrary expressions. The bug is somewhere before Rational is called because it shouldn't be passing in `1*1`.

**Comment by Mayank-gaur** [comment]
HI! I am new to this organization and I want to work on this issue. Can somebody please guide how can I get started?

**Comment by oscarbenjamin** [comment]
The OP issue seems to be fixed already:
```python
In [11]: with evaluate(False): p = Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3)))

In [12]: p
Out[12]: 
⎧  -1         for x < 2    
⎪                          
⎨2⋅x - 1  for 2 ≤ x ∧ x ≤ 3
⎪                          
⎩   1         for x > 3 
```

**Comment by faze-geek** [comment]
Is this issue handled ?Is some work( like adding tests maybe )left to be done ? 
Else I would request @oscarbenjamin to close this issue !

**Comment by faze-geek** [comment]
Yeah sure ,I'll do that and submit a closing pull request for this issue .

**Comment by faze-geek** [comment]
@oscarbenjamin
I have attempted to add test cases but can't figure out why it gives an assertion error .Being a new contributor to sympy it would be helpful if you can provide some guidance regarding this .

```
def test_issue_20715():
    from sympy.core.parameters import evaluate
    x = Symbol('x')
    with evaluate(False):        
        assert Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,StrictGreaterThan(x, 3))) == Piecewise((-1, x < 2), (2*x - 1, (2 <= x) & (x <= 3)), (1, x > 3)) 
```
Thanks

**Comment by me-t1me** [comment]
@faze-geek you have to install dev package of sympy for it.

**Comment by faze-geek** [comment]
> @faze-geek you have to install dev package of sympy for it.

Not sure what you mean by this .For contributing to SymPy ofcourse I am using the dev package .

**Comment by me-t1me** [comment]
as of version `1.10 dev` this issue is fixed @faze-geek.

**Comment by oscarbenjamin** [comment]
This was fixed in ceec8c755703cf1e9cd8d5ce61839044ee377562 from #22319

**Comment by faze-geek** [comment]
> This was fixed in [ceec8c7](https://github.com/sympy/sympy/commit/ceec8c755703cf1e9cd8d5ce61839044ee377562) from #22319
@oscarbenjamin 
Yeah can you have a look at the assert I had mentioned above ,I had tried it a while ago ! Why does it give assertion error .The fixing PR would be supported by that test and potentially then this issue can be closed .

**Comment by oscarbenjamin** [comment]
The unevaluated expressions are not exactly the same because the direction of an inequality would be flipped when evaluating:
```python
In [35]: with evaluate(False):
    ...:     p1 = Piecewise((-1,StrictLessThan(x, 2)),(2*x-1,And(LessThan(2, x), LessThan(x, 3))),(1,
    ...: StrictGreaterThan(x, 3)))
    ...:     p2 = Piecewise((-1, x < 2), (2*x - 1, (2 <= x) & (x <= 3)), (1, x > 3))
    ...: 

In [36]: p1
Out[36]: 
⎧  -1         for x < 2    
⎪                          
⎨2⋅x - 1  for 2 ≤ x ∧ x ≤ 3
⎪                          
⎩   1         for x > 3    

In [37]: p2
Out[37]: 
⎧  -1         for x < 2    
⎪                          
⎨2⋅x - 1  for x ≥ 2 ∧ x ≤ 3
⎪                          
⎩   1         for x > 3    

In [38]: p1 == p2
Out[38]: False

In [39]: p1.doit() == p2.doit()
Out[39]: True
```

**Comment by faze-geek** [comment]
Thanks now I get it , would that test be needed ?
